In [36]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
from sklearn.dummy import DummyClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.svm import LinearSVC
from sklearn.decomposition import TruncatedSVD
from sklearn.naive_bayes import MultinomialNB
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import ExtraTreesClassifier
from sklearn.ensemble import AdaBoostClassifier
from sklearn.ensemble import BaggingClassifier
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.preprocessing import OneHotEncoder
from sklearn.pipeline import Pipeline
from sklearn.metrics import roc_curve, auc, classification_report
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

In [63]:
# Read in datasets
training_data = pd.read_csv("si670_kaggle1_train.csv")
test_data = pd.read_csv("test.csv")
validation_data = pd.read_csv("si670_kaggle1_validation.csv")

In [67]:
# Split into features and labels
X_train, y_train = training_data["text"], training_data["label"]
X_val, y_val = validation_data["text"], validation_data["label"]
X_test = test_data["text"]

In [ ]:
# Preprocess data
# pipe = Pipeline([
#     ('vect', TfidfVectorizer()),
#     ('logreg', LogisticRegression())
# ])

# param_grid = [
#     {
#         'vect': [CountVectorizer(), TfidfVectorizer()],
#         'vect__ngram_range': [(1,1), (1,2)],
#         'vect__min_df': [1, 2],
#         'logreg__C': [0.01, 0.1, 1, 10],
#         'logreg__penalty': ['l2'],
#         'logreg__solver': ['lbfgs']
#     }
# ]

vectorizer = TfidfVectorizer()
X_train_tfidf = vectorizer.fit_transform(X_train)
X_val_tfidf = vectorizer.transform(X_val)
X_test_tfidf = vectorizer.transform(X_test)

KeyboardInterrupt: 

In [ ]:
# lg = LogisticRegression(penalty='l1',solver='liblinear')
# sv = SVC(kernel='linear',gamma=1.0)
# mnb = MultinomialNB()
# dtc = DecisionTreeClassifier(max_depth=5)
# knn = KNeighborsClassifier(n_neighbors=5)
# rfc = RandomForestClassifier(n_estimators=50,random_state=42)
# etc = ExtraTreesClassifier(n_estimators=50,random_state=42)
# abc = AdaBoostClassifier(n_estimators=50,random_state=42)
# bg = BaggingClassifier(n_estimators=50,random_state=42)
# gbc = GradientBoostingClassifier(n_estimators=50,random_state=42)

In [39]:
accuracy_scores = {}
f1_scores = {}

In [79]:

# pipeline = Pipeline([
#     ("tfidf", TfidfVectorizer(max_features=50000, ngram_range=(1,2))),
#     # ("svd", TruncatedSVD(n_components=100, random_state=42)),
#     ("lin_svc", LinearSVC(C=c, random_state=42))
# ])


pipeline = Pipeline([
    ("tfidf", TfidfVectorizer(max_features=50000, ngram_range=(1,2), analyzer='word',
                              min_df=2, max_df=0.9)),
    # ("svd", TruncatedSVD(n_components=100, random_state=42)),
    ("lin_svc", LinearSVC(C=1, random_state=42, class_weight='balanced'))
])
pipeline.fit(X_train, y_train)
predictions = pipeline.predict(X_val)
    # accuracy_scores[c] = accuracy_score(y_val, predictions)
    # f1_scores[c] = f1_score(y_val, predictions)

print(accuracy_score(y_val, predictions))
print(f1_score(y_val, predictions))

0.7426926327651782
0.7504994109511858


In [78]:
accuracy = accuracy_score(y_val, predictions)
f1 = f1_score(y_val, predictions)
print(f"Accuracy: {accuracy}, F1 Score: {f1}")
accuracy_scores["log_reg"] = accuracy
f1_scores["log_reg"] = f1

Accuracy: 0.7300676151570644, F1 Score: 0.711722892925646


Accuracy: 0.7300676151570644, F1 Score: 0.711722892925646

`pipeline = Pipeline([
    ("tfidf", TfidfVectorizer(max_features=50000, ngram_range=(1,2), analyzer='word',
                              min_df=2, max_df=0.9)),
    # ("svd", TruncatedSVD(n_components=100, random_state=42)),
    ("lin_svc", LinearSVC(C=1, random_state=42))
])
pipeline.fit(X_train, y_train)
predictions = pipeline.predict(X_val)`

In [ ]:
# etc.fit(X_train_tfidf, y_train)
# pred = etc.predict(X_val_tfidf)
# accuracy = accuracy_score(y_val, pred)
# f1 = f1_score(y_val, pred)

In [ ]:
final_predictions = pipeline.predict(X_test)

In [ ]:
with open("si670_kaggle1_predictions.csv", "w") as f:
    f.write("id,label\n")
    for i in range(len(final_predictions)):
        f.write(f"{X_test.index[i]},{final_predictions[i]}]")

In [23]:
# def prediction(model,X_train,X_val,y_train,y_val):
#     model.fit(X_train,y_train)
#     pred = model.predict(X_val)
#     acc_score = accuracy_score(y_val,pred)
#     f1 = f1_score(y_val, pred)
#     return acc_score,f1

# acc_scores = {}
# f1_scores = {}

# clf = knn
# acc_scores["knn"],f1_scores["knn"]= prediction(clf,X_train_tfidf,X_val_tfidf,y_train,y_val)
# print(f"Accuracy: {acc_scores['knn']}, F1 Score: {f1_scores['knn']}")

In [8]:
# def prediction(model,X_train,X_val,y_train,y_val):
#     model.fit(X_train,y_train)
#     pred = model.predict(X_val)
#     acc_score = accuracy_score(y_val,pred)
#     f1 = f1_score(y_val, pred)
#     return acc_score,f1
   
# acc_scores = {}
# f1_scores = {}
# clfs= {
#     'LR':lg,
#     'SVM':sv,
#     # 'DTC':dtc,
#     'KNN':knn,
#     # 'RFC':rfc,
#     # 'ETC':etc,
#     # 'ABC':abc,
#     # 'BG':bg,
#     # 'GBC':gbc,
# }
# for name,clf in clfs.items():
#     acc_scores[name],f1_scores[name]= prediction(clf,X_train_tfidf,X_val_tfidf,y_train,y_val)
    
#     #View those scores
#     acc_scores
#     f1_scores

In [9]:
# grid = GridSearchCV(pipe, param_grid, cv=3, scoring="f1", n_jobs=-1)

# # TODO: Fit the grid search on the training data
# grid.fit(X_train, y_train)

# # TODO: Print the best parameters and CV score
# print("Best parameters:", grid.best_params_)
# print("Best CV score:", grid.best_score_)